src/offline.ts

```diff
@@ -1,14 +1,64 @@
-import Dexie, { Table } from 'dexie';
-import { v4 } from 'uuid';
+import {
+  generateOfflineUuid,
+  getSynchronizationItems,
+  isOfflineUuid,
+  QueueItemDescriptor,
+  queueSynchronizationItem,
+  setupOfflineSync,
+  subscribeNetworkRequestFailed,
+} from '@openmrs/esm-framework';
 
-const offlineUuidPrefix = 'OFFLINE+';
+export interface QueuedEncounterRequest {
+  url: string;
+  method: string;
+  headers: Record<string, string>;
+  body: any;
+}
 
-export async function syncQueuedEncounterRequests() {
-  const db = new FormEntryDb();
-  const encounterRequests = await db.encounterRequests.toArray();
+const syncType = 'form-entry-encounters';
+
+export function setupEncounterRequestInterceptors() {
+  subscribeNetworkRequestFailed(async (data) => {
+    const createEncounterPattern = /.+\/ws\/rest\/v1\/encounter$/;
+    const updateEncounterPattern = /.+\/ws\/rest\/v1\/encounter\/(.+)/;
+
+    if (data.request.method === 'POST') {
+      if (createEncounterPattern.test(data.request.url)) {
+        const { uuid } = JSON.parse(data.request.headers['x-omrs-offline-response-body']);
+        const encounterPost = JSON.parse(data.request.body);
+        const body = { ...encounterPost, uuid };
+        addAllOfflineUuids(body);
+
+        await queueEncounterRequest({
+          url: data.request.url,
+          method: data.request.method,
+          headers: data.request.headers,
+          body,
+        });
+      } else if (updateEncounterPattern.test(data.request.url)) {
+        const uuid = updateEncounterPattern.exec(data.request.url)[1];
+        const encounterRequest = await findQueuedEncounterRequest(uuid);
+        const existingEncounter = encounterRequest.body;
+        const updatedEncounter = JSON.parse(data.request.body);
+        const newEncounter = mergeEncounterUpdate(existingEncounter, updatedEncounter);
+        encounterRequest.body = newEncounter;
+        await queueEncounterRequest(encounterRequest);
+      }
+    }
+  });
+}
 
-  for (const { id, request } of encounterRequests) {
-    const body = { ...request.body };
+async function queueEncounterRequest(item: QueuedEncounterRequest) {
+  const descriptor: QueueItemDescriptor = {
```

Comment:
I ❤️ that the new descriptors can also be used to update/replace an existing item in the sync queue!